riscof-plugin/sail_cSim/env/model_test.h

```diff
@@ -0,0 +1,60 @@
+#ifndef _COMPLIANCE_MODEL_H

```

Comment:
None of this looks specific to Sail, other than the tohost/fromhost usage that will still be common across Sail and QEMU. A basic common version of this belongs in riscof, not duplicated in every project.

Reply:
The point here is exactly to have it in tied to every model. Just because two models share this doesn't mean that all models share this. The majority will not. Sail and QEMU may - but there are lot more models out there than just those two, and we need to make it work for all of them.

Reply:
HTIF is a de-facto standard. Other platform-specific mechanisms are not so. That's the difference.

Reply:
Yes, but our view is that we need to work with everybody, so anything that might be different has to be supplied - and that includes things that aren't different.  I'm not sure de-facto standard quite meets that bar, but we'll need to discuss that.
I think what you're asking for here (and in other places) is that there be a default that can be overwritten, which has a different set of issues. More discussion...

Reply:
I think there are two things:
1. Defaults, like linking at 0x80000000 as used by almost all existing platforms
2. Optional common code for de-facto standards, but would still be opt-in, just there for you to use so you can say "I use this common standard method" without having to write your own version or make another copy of the same thing
I don't know if I'd class the HTIF in 1 or 2; in a sense it's 2, but because it's assumed by the existing ISA tests it's effectively an enforced testing standard at this point so may make sense even as a default.

Reply:
Oddly, you first example is not shared by (my company's) platform. But I don't believe that statement even if we did. It maybe be true for the platforms you've looked at - but there are hundreds of vendors out there building harts with thousands of transistors, and with billions of transistors. And: we are not testing platforms, we are testing harts. We don't know what will be built into PMAs that we can't control and have no visibility into. We have no idea what will be a de-facto standard and have neither resources nor desire to survey them to see if there is one (or two or three) that appear to be at this very moment. That doesn't scale. What scales is letting every vendor specify their own. They are specifying their RTL, they can damn will specify a linker script that works with their RTL. If you want to create some common code and add it to the /doc directory as an example of a link script or whatever, go right ahead, we'd probably accept it. Otherwise, no. It's the vendor's responsibilty.